File: 2024/day02/main.lua
local reports = {}

for line in io.lines('input') do
  local levels = {}
  for level in line:gmatch('%S+') do
    table.insert(levels, tonumber(level))
  end
  table.insert(reports, levels)
end

local function safe(report)
  local d0 = report[2] - report[1]
  for i = 2, #report do
    local d = report[i] - report[i - 1]
    if d == 0 or math.abs(d) > 3 or d * d0 < 0 then
      return false
    end
  end
  return true
end

-- Part 1: 591
local sum = 0
for _, r in ipairs(reports) do
  if safe(r) then
    sum = sum + 1
  end
end
print(sum)

-- Part 2: 621
local sum2 = 0
for _, r in ipairs(reports) do
  if safe(r) then
    sum2 = sum2 + 1
  else
    for i = 1, #r do
      local modified = {}
      for j, v in ipairs(r) do
        modified[j] = v
      end
      table.remove(modified, i)
      if safe(modified) then
        sum2 = sum2 + 1
        break
      end
    end
  end
end
print(sum2)

File: 2024/day02/main.py
reports = []
lines = open('input').readlines()
for line in lines:
	levels = list(map(int, line.split()))
	reports.append(levels)


def safe(report):
	d0 = report[1] - report[0]
	for i in range(1, len(report)):
		d = report[i] - report[i - 1]
		if d == 0 or abs(d) > 3 or d * d0 < 0:
			return False
	return True


# Part 1: 591
print(sum(1 for r in reports if safe(r)))


# Part 2: 621
sum = 0
for r in reports:
	if safe(r):
		sum += 1
		continue
	for i in range(len(r)):
		if safe(r[:i] + r[i + 1 :]):
			sum += 1
			break
print(sum)
